cpp/test/preprocess_utils.cu

```diff
@@ -0,0 +1,283 @@
+/*
+ * Copyright (c) 2024, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include <raft/core/device_coo_matrix.hpp>
+#include <raft/core/resource/cuda_stream.hpp>
+#include <raft/core/resource/thrust_policy.hpp>
+#include <raft/linalg/map_reduce.cuh>
+#include <raft/random/rmat_rectangular_generator.cuh>
+#include <raft/random/rng.cuh>
+#include <raft/sparse/convert/coo.cuh>
+#include <raft/sparse/matrix/preprocessing.cuh>
+#include <raft/sparse/neighbors/cross_component_nn.cuh>
+#include <raft/sparse/op/filter.cuh>
+#include <raft/sparse/op/sort.cuh>
+
+namespace raft::util {
+
+template <typename T1, typename T2>
+struct check_zeroes {
+  float __device__ operator()(const T1& value, const T2& idx)
+  {
+    if (value == 0) {
+      return 0.f;
+    } else {
+      return 1.f;
+    }
+  }
+};
+
+template <typename T1, typename T2>
+void preproc_kernel(raft::resources& handle,
+                    raft::host_vector_view<T1> h_rows,
+                    raft::host_vector_view<T1> h_cols,
+                    raft::host_vector_view<T2> h_elems,
+                    raft::device_vector_view<T2> results,
+                    int num_rows,
+                    int num_cols,
+                    bool tf_idf)
+{
+  cudaStream_t stream = raft::resource::get_cuda_stream(handle);
+  int rows_size       = h_rows.size();
+  int cols_size       = h_cols.size();
+  int elements_size   = h_elems.size();
+  auto device_matrix  = raft::make_device_matrix<T2, int64_t>(handle, num_rows, num_cols);
+  raft::matrix::fill<float>(handle, device_matrix.view(), 0.0f);
+  auto host_matrix = raft::make_host_matrix<T2, int64_t>(handle, num_rows, num_cols);
+  raft::copy(host_matrix.data_handle(), device_matrix.data_handle(), device_matrix.size(), stream);
+
+  for (int i = 0; i < elements_size; i++) {
```

Comment:
Unclear on the primary objective of this logic, but just a heads-up: unless you explicitly sync on the stream before this line, we can't assume `host_matrix` will have the same value as `device_matrix`.

Reply:
What is happening here is that I am loading the incoming host data into a dense matrix. So we represent the COO arrays as a dense matrix and then I am copying that dense matrix from host memory to GPU memory. Before that line I am expecting that both the host and device matrices are zero filled. I did it this way to use raft APIs as much as possible. For loop on host memory did not seem like the most efficient way to fill a matrix.